docs/cookbook/creating-webhooks.md

```diff
@@ -0,0 +1,105 @@
+        // and whether it is being published for the first time
+        // https://v3.docs.apostrophecms.org/reference/server-events.html#afterpublish
+        if (!data.firstTime) {
+          return;
+        }
+        // The url added here should be changed to the desired endpoint
+        const url = 'https://enr061vpkj3d.x.pipedream.net/';
+        const options = {
+          headers: {
+            // Header information will depend on the endpoint you are sending to
+            'Content-Type': 'application/json'
+          },
+          body: {
+            docTitle: data.published.title,
+            docId: data.published._id
+          }
+        };
+        const response = await self.apos.http.post(url, options);
+        // the response will depend on the endpoint you are sending to
+        // Potentially add code to handle errors coming back
+        // like retrying if the response isn't successful
+      }
+    }
+  };
+}
+```
+  <template v-slot:caption>
+    /modules/article/index.js
+    </template>
+
+</AposCodeBlock>
+
+In this block of code, we are using the `handlers(self)` module configuration function. We could use any server event, but in this case we are using the `afterPublication` event that is emitted by our custom article piece-type module. This server event delivers two parameters, `req` and `data`. The `data` parameter contains information about the document being published including the content and whether this is the first time it is being published. In this case, we are adding an early return if this isn't the first time the document is being published.
+
+The next section of code is involved in setting up the HTTP `POST` request and will depend heavily on what the desired endpoint requires, and what content you want sent to the endpoint. In this example, we are adding the piece `title` and `_id`.
+
+Finally, we are using the `post` method of the [`@apostrophecms/http` module](https://v3.docs.apostrophecms.org/reference/modules/http.html#async-post-url-options) to send our data to the endpoint. Depending on the endpoint, the returned response might be as simple as a `200` success or a `400` failed response, or might contain additional data. That response can be handled to retry a failed response or in some way log the returned data.
+
+## Setting up a webhook endpoint
+
+Your project may require a webhook endpoint for interfacing with a payment or subscription service. This can be accomplished by using the Apostrophe [`apiRoutes(self)` module configuration](https://v3.docs.apostrophecms.org/reference/module-api/module-overview.html#apiroutes-self) function. This allows you to easily set up an endpoint for any HTTP request method.
+
+<AposCodeBlock>
+
+```javascript
+options: {
+  csrfExceptions: [ '/webhooks' ]
+},
+apiRoutes(self) {
+  return {
+    post: {
+      // change function name to match the desired endpoint URL
+      '/webhooks': async function(req) {
```

Comment:
I would suggest not rooting this with a `/` in the example, but instead explaining it'll automatically be `/api/v1/your-module-name/your-method-name` for tidy namespacing, unless you do start it with a `/`. Usually services to which you provide a webhook like this allow you to paste the URL you want.

Reply:
If you just put the function name in `csrfExceptions` (without the `/`) it will do the same math to figure out the URL.

I can see how this is kind of a lot when I spell it all out. Maybe you're right to just use a rooted example with `/`.

Reply:
Yes.